### ISSUE TITLE
Retention of gradient penalty Graph

### ISSUE THREAD

**Issue by Baukebrenninkmeijer** [sdv-dev/CTGAN#37]
Hi guys, 

What is the role of retaining the graph of the gradient penalty after doing the backward pass? I couldn't find any other usage after that so I was wondering if this was still doing something? Thanks!

https://github.com/sdv-dev/CTGAN/blob/f71c8fa33257dca0e73282a7984742e3d1f97267/ctgan/synthesizer.py#L190

**Comment by leix28** [comment]
Hi @Baukebrenninkmeijer 

In PyTorch, the computing graph will be deleted after running backward. But here the computing graph is still needed for another backward pass (loss_d.backward). retain_graph prevents pytorch from deleting the computing graph.

**Comment by Baukebrenninkmeijer** [comment]
@leix28 Hmm, very interesting. It was my understanding that the gradients are stored in the tensors and you don't need to store the graph. And unless you do the optimizer step, the gradients will be stored there and you can manipulate and update them. 

In other words, calling `.backward()` three times will add the gradients of those three backward passes after which you can do the optimizer step. 

Additionally, previously the gradient penalty was an update to the loss. However, now you are doing an actual backward pass with it. What effect does this have and why was it necessary?

**Comment by leix28** [comment]
> Hmm, very interesting. It was my understanding that the gradients are stored in the tensors and you don't need to store the graph. And unless you do the optimizer step, the gradients will be stored there and you can manipulate and update them.

According to my understanding, gradients are computed when `backward` is called. And one side effect is it deletes the computation graph so that additional `backward` calls can not be executed. 

Here is an equivalent implementation:
```
loss_sum = loss_d + pen

optimizerD.zero_grad()
loss_sum.backward()
optimizerD.step()
```

**Comment by Baukebrenninkmeijer** [comment]
> According to my understanding, gradients are computed when `backward` is called. And one side effect is it deletes the computation graph so that additional `backward` calls can not be executed.

After a bit more digging: if you do not have retain_graph=True, the computational graph of that specific loss is freed from the buffer. However, the computational graph of other losses is not. 
This means you cannot call `.backward()` on a single loss twice, but you can call it on separate losses. I did a short experiment where you can see they are all equivalent to each other. 

Using the following: 
```python
init_params = torch.randn(3).requires_grad_()
def f(x, params): return params[0]*x**2 + params[1]*x + params[2]

data = torch.tensor(list(range(-10, 10)))
targets = (f(data, [1, 0, 0]) + torch.randn(20)*2).detach()

loss_func1 = torch.nn.functional.mse_loss
loss_func2 = torch.nn.functional.l1_loss
```
We get:
![image](https://user-images.githubusercontent.com/9077462/79971514-b1b9ee00-8494-11ea-9623-783fb2d09c1f.png)

So in that sense I don't think it is required in your situation. Let me know what you think!

**Comment by leix28** [comment]
Sure. Maybe in this case, retain_graph is not required, but it either way gives the same output : )
